endpoint.go

```diff
@@ -0,0 +1,126 @@
+package ngrok
+
+import (
+	"context"
+	"crypto/tls"
+	"net/url"
+	"sync"
+)
+
+// Endpoint is the interface implemented by both EndpointListener and
+// EndpointForwarder.
+type Endpoint interface {
+	// Agent returns the Agent that created this Endpoint.
+	Agent() Agent
+
+	// PoolingEnabled returns whether the endpoint supports pooling set by WithPoolingEnabled.
+	PoolingEnabled() bool
+
+	// Bindings returns the endpoint's bindings set by WithBindings
+	Bindings() []string
+
+	// Close() is equivalent to for CloseWithContext(context.Background())
+	Close() error
+
+	// CloseWithContext closes the endpoint with the provided context.
+	CloseWithContext(context.Context) error
+
+	// Description returns the endpoint's human-readable description set by WithDescription.
+	Description() string
+
+	// Done returns a channel that is closed when the endpoint stops.
+	Done() <-chan struct{}
+
+	// ID returns the unique endpoint identifier assigned by the ngrok cloud service.
+	ID() string
+
+	// Metadata returns the endpoint's opaque user-defined metadata set by WithMetadata.
+	Metadata() string
+
+	// Protocol is sugar for URL().Scheme
+	Protocol() string
+
+	// ProxyProtocol returns the proxy protocol setting for the endpoint.
+	ProxyProtocol() string
```

Comment:
We call `ProxyProtocol` in the configuration file `upstream.proxy_protocol` wondering if we should change this in either the configuration file or in here.